src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1029 @@
+	int const char_max_row = cliprect.max_y/14;
+	bool const blink_strobe = m_frame_counter&0x08;
+	bool const cursor_blink_strobe = blink_strobe && !m_speed_check;
+
+	for(int char_row_nr = char_min_row; char_row_nr <= char_max_row; char_row_nr++)
+	{
+		// This is at the start of a char row
+		int const chr_y_pos = char_row_nr*14;
+		int attribute = get_attribute(char_row_nr, char_min_col);
+
+		for (int char_col_nr = char_min_col; char_col_nr <= char_max_col; char_col_nr++)
+		{
+			// This is at the start of a single character
+			int const chr_x_pos = char_col_nr*9;
+			int const in_line_char_nr = char_col_nr - 4;
+			int const vram_address = get_ram_addr(char_row_nr, in_line_char_nr);
+			bool const is_cursor = (m_cursor_col == in_line_char_nr && m_cursor_row == char_row_nr);
+
+			for(int line = 0; line < 14; line++)
+			{
+				// This is at the start of a 9-dot raster-line segment in a single character
+				uint8_t data = 0x00;
+				int dot_nr8 = 0;
+				int extra_intensity = 1;
+				if(in_line_char_nr >= 0 && in_line_char_nr < 80)
+				{
+					bool const draw_cursor = (is_cursor && (line == 12 || block_cursor) && (!blinking_cursor || cursor_blink_strobe));
+					int const vram_data = m_vram[vram_address];
+
+					// Pattern generator & character blanking
+					if(m_video_enable && (vram_data&0x60 || !blank_ctrl_chars) && !(vram_data&0x80 && blank_attr_chars))
+					{
+						data = m_font[((vram_data&0x7f)<<4) + line];
+						if(extend_dot7)
+						{
+							dot_nr8 = data&0x01;
+						}
+					}
+
+					// Pattern modifiers
+					bool underline_mode = m_dsw_u61->read()&0x008;
+					if(underline_mode)
+					{
+						// Note: The underline in underline-mode is not affected by blanking!
+						if((vram_data&0x80) && line == 13)
+						{
+							data = 0xff;
+							dot_nr8 = 0x01;
+						}
+					}
+					else if(m_video_enable) // attribute mode
+					{
+						if(vram_data&0x80)
+						{
+							attribute = (vram_data&0x70)>>4;
+						}
+						else
+						{
+							if(attribute == 2)
+							{
+								// Low intensity
+								extra_intensity = 0;
+							}
+							else if(attribute == 3 && blink_strobe && !draw_cursor)
+							{
+								// Blink
+								data = 0x00;
+								dot_nr8 = 0x00;
+							}
+							else if(attribute == 4)
+							{
+								// Inverse
+								data = ~data;
+								dot_nr8 = ~dot_nr8;
+							}
+							else if(attribute == 5 && line == 13)
+							{
+								// Underline
+								data = 0xff;
+								dot_nr8 = 0x01;
+							}
+							else if(attribute == 6)
+							{
+								// Invissible
+								data = 0x00;
+								dot_nr8 = 0x00;
+							}
+						}
+					}
+
+					if(draw_cursor)
+					{
+						data = ~data;
+						dot_nr8 = ~dot_nr8;
+					}
+				}
+				if(full_inverse_video)
+				{
+					data = ~data;
+					dot_nr8 = ~dot_nr8;
+				}
+				for(int dot = 0; dot < 8; dot++)
+				{
+					bitmap.pix(chr_y_pos + line, chr_x_pos + dot) = palette[((data>>(7-dot))&0x01)<<extra_intensity];
+				}
```

Comment:
Please use the `BIT` helper for extracting individual bits – `BIT(data, 7 - dot)` is a lot easier to read than `((data>>(7-dot))&0x01)`.

Reply:
This assumes BIT returns either 0 or 1, and will never by itself cast to bool in any foreseeable future. I was not able to find the definition used for this macro, but I can change it if we know for sure we have this under control.

Reply:
BIT is defined here: https://github.com/mamedev/mame/blob/master/src/lib/util/coretmpl.h#L584, it should always be a 0 or 1.

Reply:
Ok, adjusted.